src/drivers/cyphal/Cyphal.cpp

```diff
@@ -305,101 +228,62 @@ void UavcanNode::Run()
 	pthread_mutex_unlock(&_node_mutex);
 }
 
-void UavcanNode::transmit()
+template <typename As, typename F>
+static void traverseTree(const CanardTreeNode *const root, const F &op)  // NOLINT this recursion is tightly bounded
 {
-	// Look at the top of the TX queue.
-	for (const CanardFrame *txf = nullptr; (txf = canardTxPeek(&_canard_instance)) != nullptr;) {
-		// Attempt transmission only if the frame is not yet timed out while waiting in the TX queue.
-		// Otherwise just drop it and move on to the next one.
-		const hrt_abstime now = hrt_absolute_time();
-
-		if (txf->timestamp_usec == 0 || txf->timestamp_usec > now) {
-			// Send the frame. Redundant interfaces may be used here.
-			const int tx_res = _can_interface->transmit(*txf);
-
-			if (tx_res < 0) {
-				// Failure - drop the frame and report
-				canardTxPop(&_canard_instance);
-
-				// Deallocate the dynamic memory afterwards.
-				_canard_instance.memory_free(&_canard_instance, (CanardFrame *)txf);
-				PX4_ERR("Transmit error %d, frame dropped, errno '%s'", tx_res, strerror(errno));
-
-			} else if (tx_res > 0) {
-				// Success - just drop the frame
-				canardTxPop(&_canard_instance);
-
-				// Deallocate the dynamic memory afterwards.
-				_canard_instance.memory_free(&_canard_instance, (CanardFrame *)txf);
-
-			} else {
-				// Timeout - just exit and try again later
-				break;
-			}
-
-		} else if (txf->timestamp_usec <= now) {
-			// Transmission timed out -- remove from queue and deallocate its memory
-			canardTxPop(&_canard_instance);
-
-			_canard_instance.memory_free(&_canard_instance, (CanardFrame *)txf);
-		}
+	if (root != nullptr) {
+		traverseTree<As, F>(root->lr[0], op);
+		op(static_cast<const As *>(static_cast<const void *>(root)));
+		traverseTree<As, F>(root->lr[1], op);
 	}
 }
 
-void UavcanNode::print_info()
+void CyphalNode::print_info()
 {
 	pthread_mutex_lock(&_node_mutex);
 
 	perf_print_counter(_cycle_perf);
 	perf_print_counter(_interval_perf);
 
-	O1HeapDiagnostics heap_diagnostics = o1heapGetDiagnostics(uavcan_allocator);
+	/*O1HeapDiagnostics heap_diagnostics = o1heapGetDiagnostics(uavcan_allocator);
```

Comment:
Why comment this out?

Reply:
Oops that went wrong migrating O1Heap to the CanardHandle class, I've created a new PR to fix this.